docs/proposals/kubelet-eviction.md

```diff
@@ -0,0 +1,259 @@
+other resources like disk.  The proposal focuses on a simple default eviction strategy
+intended to cover the broadest class of user workloads.
+
+## Eviction Signals
+
+The `kubelet` will support the ability to trigger eviction decisions on the following signals.
+
+| Eviction Signal  | Description                                                                     |
+|------------------|---------------------------------------------------------------------------------|
+| memory.available | memory.available := node.status.capacity[memory] - node.stats.memory.workingSet |
+
+## Eviction Thresholds
+
+The `kubelet` will support the ability to specify eviction thresholds.
+
+An eviction threshold is of the following form:
+
+`<eviction-signal><operator><quantity>`
+
+* valid `eviction-signal` tokens as defined above.
+* valid `operator` tokens are `<`
+* valid `quantity` tokens must match the quantity representation used by Kubernetes
+
+If threhold criteria are met, the `kubelet` will take pro-active action to attempt
+to reclaim the starved compute resource associated with the eviction signal.
+
+The `kubelet` will support soft and hard eviction thresholds.
+
+### Soft Eviction Thresholds
+
+A soft eviction threshold pairs an eviction threshold with a required
+administrator specified grace period.  No action is taken by the `kubelet`
+to reclaim resources associated with the eviction signal until that grace
+period has been exceeded.  If no grace period is provided, the `kubelet` will
+error on startup.
+
+In addition, if a soft eviction threshold has been met, an operator can
+specify a maximum allowed pod termination grace period to use when evicting
+pods from the node.  If not specified, the `kubelet` will kill pods immediately
+with no grace period.
+
+To configure soft eviction thresholds, the following flags will be supported:
+
+```
+--eviction-soft="": A set of eviction thresholds (e.g. memory.available<1.5Gi) that if met over a corresponding grace period would trigger a pod eviction.
+--eviction-soft-grace-period="": A set of eviction grace periods (e.g. memory.available=1m30s) that correspond to how long a soft eviction threshold must hold before triggering a pod eviction.
+--eviction-soft-max-pod-termination-grace-period-seconds="0": Maximum allowed pod termination grace period to use when evicting pods from the node in response to a soft eviction threshold being met.
+```
+
+### Hard Eviction Thresholds
+
+A hard eviction threshold has no grace period, and if observed, the `kubelet`
+will take immediate action to reclaim the associated starved resource.  If a
+hard eviction threshold is met, the `kubelet` will kill the pod immediately
+and not respect its termination grace period.
+
+To configure hard eviction thresholds, the following flag will be supported:
+
+```
+--eviction-hard="": A set of eviction thresholds (e.g. memory.available<1Gi) that if met would trigger a pod eviction.
```

Comment:
>How do you envision admins setting this flag?

Based on workload and experience.  Average number of pods per node, etc.  Many clusters will have utilization metrics that they have seen and will have experienced spikes in utilization that induced OOM.  If utilization is a top priority, then I suspect the flag would be set uniformly across the cluster with a value that satisfies their target balanced with their willingness to accept OOM, speed to respond, and ability to procure new nodes.

Many operators are not pushing utilization to huge numbers and are interested in node stability and just the consistent control plane afforded by Kubernetes/OpenShift.  In those cases, I suspect they would set eviction thresholds to happen earlier since they probably have slack on other nodes in the cluster and are comfortable or desire lower utilization for their workload.

For the vast majority though, I would imagine folks will follow this pattern:

* operators are not seeing OOMs, so they set nothing
* cluster gets popular, users do more, start to experience OOM on nodes regularly
* operator files a bug with us and buys more nodes temporarily
* operators notice users are not doing the right thing with their requests, or using best effort pods to run dbs, etc.
* operators decide to experiment on a month-by-month basis to set eviction flags until their incidence of oom observed reaches satisfactory outcomes.
* users continue to do unexpected things like run dbs in besteffort pods

>Is a homogeneous cluster assumed here?

No.  Tailoring values based on node profiles is probably something I would expect Ansible to do today probably similar to how we can specify node labels.